File: modules/04_privacy_preserving_ai/code_samples/ios_hybrid_fallback/HybridClassifier.swift
// Implements local-first, optional cloud fallback:
struct HybridClassifier {
    private let coreModel = ImageClassifier()

    func classify(
        _ image: UIImage,
        allowCloudFallback: Bool
    ) async -> (label: String, confidence: Float) {

        let local = coreModel.classify(image)

        if local.confidence >= 0.8 {
            return local
        }

        guard allowCloudFallback else {
            return ("Unsure", local.confidence)
        }

        // stub for demo – in a real app this calls your server/LLM
        let cloud = await cloudPredict(image)
        return cloud
    }

    private func cloudPredict(_ image: UIImage) async -> (String, Float) {
        // demo only
        return ("cloud_label", 0.9)
    }
}

File: modules/04_privacy_preserving_ai/code_samples/android_hybrid_fallback/HybridClassifier.kt
class HybridClassifier(
    private val localClassifier: Classifier,
    private val allowCloudFallback: Boolean
) {

    suspend fun classify(bitmap: Bitmap): ClassificationResult {
        val local = localClassifier.classify(bitmap)

        if (local.confidence >= 0.8f) {
            return ClassificationResult(local.label, local.confidence)
        }

        if (!allowCloudFallback) {
            return ClassificationResult("unsure", local.confidence)
        }

        val cloud = cloudPredict(bitmap) // stub
        return cloud
    }

    private suspend fun cloudPredict(bitmap: Bitmap): ClassificationResult {
        // demo only – in reality, upload a derived representation or compressed image
        return ClassificationResult("cloud_label", 0.9f)
    }
}
